clippy_lints/src/redundant_clone.rs

```diff
@@ -70,317 +68,724 @@ impl<'tcx> LateLintPass<'tcx> for RedundantClone {
+                    .local_data
+                    .as_ref()
+                    .assert_crate_local()
+                    .lint_root;
+                span_lint_hir(cx, REDUNDANT_CLONE, node, term.source_info.span, "redundant clone");
             }
+        }
+    }
+}
 
-            let (fn_def_id, arg, arg_ty, clone_ret) =
-                unwrap_or_continue!(is_call_with_ref_arg(cx, mir, &terminator.kind));
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Candidate {
+    Lint,
+    NoLint,
+}
 
-            let from_borrow = match_def_path(cx, fn_def_id, &paths::CLONE_TRAIT_METHOD)
-                || match_def_path(cx, fn_def_id, &paths::TO_OWNED_METHOD)
-                || (match_def_path(cx, fn_def_id, &paths::TO_STRING_METHOD)
-                    && is_type_lang_item(cx, arg_ty, LangItem::String));
+#[derive(Debug, Clone)]
+struct Value {
+    last_modified: Option<BasicBlock>,
+    lint_unread: HybridBitSet<BasicBlock>,
+}
 
-            let from_deref = !from_borrow
-                && (match_def_path(cx, fn_def_id, &paths::PATH_TO_PATH_BUF)
-                    || match_def_path(cx, fn_def_id, &paths::OS_STR_TO_OS_STRING));
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+struct Candidates(FxHashMap<BasicBlock, Candidate>);
+impl Candidates {
+    fn set_possible(&mut self, blocks: &HybridBitSet<BasicBlock>) {
+        for block in blocks.iter() {
+            match self.0.entry(block) {
+                Entry::Vacant(e) => {
+                    e.insert(Candidate::Lint);
+                },
+                Entry::Occupied(_) => {},
+            }
+        }
+    }
 
-            if !from_borrow && !from_deref {
+    fn remove(&mut self, blocks: &HybridBitSet<BasicBlock>) {
+        for block in blocks.iter() {
+            match self.0.entry(block) {
+                Entry::Vacant(e) => {
+                    e.insert(Candidate::NoLint);
+                },
+                Entry::Occupied(mut e) => {
+                    *e.get_mut() = Candidate::NoLint;
+                },
+            }
+        }
+    }
+}
+impl JoinSemiLattice for Candidates {
+    fn join(&mut self, other: &Self) -> bool {
+        let mut changed = false;
+        for (&block, candidate) in &other.0 {
+            match self.0.entry(block) {
+                Entry::Vacant(e) => {
+                    changed = true;
+                    e.insert(*candidate);
+                },
+                Entry::Occupied(mut e) => match (e.get(), candidate) {
+                    (Candidate::Lint, Candidate::NoLint) => {
+                        *e.get_mut() = Candidate::NoLint;
+                        changed = true;
+                    },
+                    _ => {},
+                },
+            }
+        }
+        changed
+    }
+}
+
+#[derive(Default, Debug, Clone)]
+struct CloneDomain {
+    links: FxHashSet<(Local, Local, BasicBlock)>,
+    values: FxHashMap<Local, Value>,
+    candidates: Candidates,
```

Comment:
This field should not be part of the analysis domain, but computed in post-processing.

Reply:
I think that would work.